contrib/rust/src/python/pants/contrib/rust/tasks/cargo_build.py

```diff
@@ -0,0 +1,355 @@
+    register(
+      '--cargo-opt',
+      type=list,
+      default=[],
+      help='Append these options to the cargo command line.')
+
+  @classmethod
+  def prepare(cls, options, round_manager):
+    super(Build, cls).prepare(options, round_manager)
+    round_manager.require_data('cargo_env')
+
+  @property
+  def cache_target_dirs(self):
+    return True
+
+  @classmethod
+  def product_types(cls):
+    return ['rust_libs', 'rust_bins', 'rust_tests']
+
+  @classmethod
+  def supports_passthru_args(cls):
+    return True
+
+  def execute(self):
+    self.prepare_task()
+    self.prepare_cargo_targets()
+
+    cargo_targets = self.context.build_graph.targets(self.is_cargo_base_target)
+    self.build_targets(cargo_targets)
+
+  def prepare_task(self):
+    self.context.products.safe_create_data('rust_libs', lambda: {})
+    self.context.products.safe_create_data('rust_bins', lambda: {})
+    self.context.products.safe_create_data('rust_tests', lambda: {})
+    self._libraries_dir_path = self.create_libraries_dir()
+
+  def create_libraries_dir(self):
+    libraries_dir_path = os.path.join(self.versioned_workdir, 'deps')
+    safe_mkdir(libraries_dir_path)
+    return libraries_dir_path
+
+  def prepare_cargo_targets(self):
+    cargo_targets = self.get_targets(self.is_cargo_workspace)
+    for target in cargo_targets:
+      cargo_build_plan = self.get_cargo_build_plan(target)
+      target_definitions = self.get_target_definitions_out_of_cargo_build_plan(cargo_build_plan)
+      self.generate_targets(target_definitions, target)
+
+  def get_cargo_build_plan(self, target):
+    with self.context.new_workunit(name='cargo-build-plan',
+                                   labels=[WorkUnitLabel.COMPILER]) as workunit:
+      abs_manifest_path = os.path.join(get_buildroot(), target.manifest, self.manifest_name())
+
+      self.context.log.info(
+        'Getting cargo build plan for manifest: {0}\nAdditional cargo options: {1}'.format(
+          abs_manifest_path, self.get_options().cargo_opt))
+
+      cmd = ['cargo', 'build',
+             '--manifest-path', abs_manifest_path,
+             '--build-plan', '-Z', 'unstable-options']
+
+      if self.include_compiling_tests():
+        cmd.extend(['--tests'])
+
+      cmd.extend(self.get_options().cargo_opt)
+
+      env = {
+        'CARGO_HOME': (self.context.products.get_data('cargo_env')['CARGO_HOME'], False),
+        'PATH': (self.context.products.get_data('cargo_env')['PATH'], True)
+      }
+
+      std_output = self.run_command_and_get_output(cmd, target.toolchain, env, workunit)
+      cargo_build_plan = json.loads(std_output)
+
+      if workunit.outcome() != WorkUnit.SUCCESS:
+        self.context.log.error(workunit.outcome_string(workunit.outcome()))
+      else:
+        self.context.log.info(workunit.outcome_string(workunit.outcome()))
+
+    return cargo_build_plan
+
+  def include_compiling_tests(self):
+    return 'test' in self.context.requested_goals
```

Comment:
There is another recommended way to accomplish this, which is by using "optional" product dependencies via `is_required_data` and `require_data`. In this case, that would look like a call to `self.products.is_required_data('rust_tests')`.

More importantly though: it's important to ensure that whether or not you `include_compiling_tests` ends up in the cache key for the run (an annoying weakness of the pants `v1` task model). In this case, I think that you might need to do that by adding a `FingerprintStrategy` that includes whether `--tests` should be passed in the fingerprint for a target (an [example strategy](https://github.com/pantsbuild/pants/blob/master/contrib/go/src/python/pants/contrib/go/tasks/go_binary_fingerprint_strategy.py) that does something similar]). You would then pass the `FingerprintStrategy` to `self.invalidated`.